### PR TITLE
Add wrapper which will allow running o/k tests as external binary in origin

### PR DESCRIPTION
**Pull request by soltysh**
This goes in pair with https://github.com/openshift/origin/pull/27570

Write-up in https://github.com/openshift/enhancements/pull/1291

### CODE REVIEW

File: openshift-hack/e2e/annotate/rules.go [hunk 1]
```diff
@@ -111,7 +110,40 @@ var (
 			`Netpol \[LinuxOnly\] NetworkPolicy between server and client using UDP should enforce policy based on Ports`,
 			`Netpol \[LinuxOnly\] NetworkPolicy between server and client using UDP should enforce policy to allow traffic only from a pod in a different namespace based on PodSelector and NamespaceSelector`,
 
+			// The new NetworkPolicy test suite is extremely resource
+			// intensive and causes itself and other concurrently-running
+			// tests to be flaky.
+			// https://bugzilla.redhat.com/show_bug.cgi?id=1980141
+			`\[sig-network\] Netpol `,
```
Comment (ingvagabund) [inline openshift-hack/e2e/annotate/rules.go:117]: These are new tests? Or, the tests are expected to be prefixed with [sig-...] from now on to be disabled? Or, these will appear after the 1.27 rebase lands?

**Comment by ingvagabund** [comment]
/lgtm
The PR introduces a new command which is better to test up-front rather than provide fully reviewed code. Some of the code is copy-pasted from origin. Changes under `openshift-hack/e2e/annotate` looks ok.
/hold
Feel free to unhold. I dropped few comments. Mostly for me to better understand the flow.

File: openshift-hack/cmd/k8s-tests/provider.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+	// this appears to inexplicably auto-register global flags.
+	_ "k8s.io/kubernetes/test/e2e/storage/drivers"
+
+	// these are loading important global flags that we need to get and set
+	_ "k8s.io/kubernetes/test/e2e"
+	_ "k8s.io/kubernetes/test/e2e/lifecycle"
+)
+
+// copied directly from github.com/openshift/origin/test/extended/util/cluster/cluster.go
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/cmd/k8s-tests/provider.go:32]: This was about these 2 files:

File: openshift-hack/cmd/k8s-tests/provider.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+func initializeTestFramework(provider string) error {
+	providerInfo := &ClusterConfiguration{}
+	if err := json.Unmarshal([]byte(provider), &providerInfo); err != nil {
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/cmd/k8s-tests/provider.go:35]: providerInfo in origin is populated here: https://github.com/openshift/origin/blob/bfcc30fde82a91fab1c881a836e4ad8e63c72fb5/cmd/openshift-tests/provider.go#L78, so it's done on every test invocation. 

tl;dr 

When you invoke `openshift-tests run openshift/conformance/parallel` it creates a list of tests and runs one by one invoking `openshift-tests run-test`. The above `decodeProvider` is responsible to read what the cluster looks like during that `run-test` invocation. Since we don't want to rely on that mechanism in here, we need to inject that value through env variable, and then decode it and use it.

[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/cmd/k8s-tests/provider.go:35]: I'm updating https://github.com/openshift/enhancements/pull/1291 I'll add more details how this works there.

File: openshift-hack/cmd/k8s-tests/provider.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+		return err
+	}
+
+	if ad := os.Getenv("ARTIFACT_DIR"); len(strings.TrimSpace(ad)) == 0 {
+		os.Setenv("ARTIFACT_DIR", filepath.Join(os.TempDir(), "artifacts"))
+	}
+
+	context.DeleteNamespace = os.Getenv("DELETE_NAMESPACE") != "false"
+	context.VerifyServiceAccount = true
+	testfiles.AddFileSource(e2etestingmanifests.GetE2ETestingManifestsFS())
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/cmd/k8s-tests/provider.go:80]: `k8s-tests run-test` is invoked once per tests, see my explanation above, `k8s-tests run-test` is identical to `openshift-tests run-test` it only run one test only.

File: openshift-hack/cmd/k8s-tests/runtest.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	suiteConfig.Timeout = 24 * time.Hour
+	reporterConfig.NoColor = true
+
+	ginkgo.SetReporterConfig(reporterConfig)
+	ginkgo.GetSuite().RunSpec(test.spec, ginkgo.Labels{}, "", ginkgo.GetFailer(), ginkgo.GetWriter(), suiteConfig)
+
+	var summary types.SpecReport
+	for _, report := range ginkgo.GetSuite().GetReport().SpecReports {
+		if report.NumAttempts > 0 {
+			summary = report
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/cmd/k8s-tests/runtest.go:74]: That's a copy of the mechanism as was in origin, but that's a good question, I'll make sure to look more into it and fix it.

File: openshift-hack/cmd/k8s-tests/provider.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ingvagabund) [reply openshift-hack/cmd/k8s-tests/provider.go:80]: I was thinking of a case when in the future `initializeTestFramework` gets accidentally invoked twice.

File: openshift-hack/e2e/annotate/annotate.go [hunk 6]
```diff
@@ -68,8 +70,11 @@ func Run() {
 	}
 
 	var pairs []string
-	for from, to := range generator.output {
-		pairs = append(pairs, fmt.Sprintf("%q:\n%q,", from, to))
+	for testName, labels := range generator.output {
+		if filter(fmt.Sprintf("%s%s", testName, labels)) {
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/e2e/annotate/annotate.go:74]: Probably, but origin will directly use the annotation mechanism, without modifying it.

File: openshift-hack/e2e/annotate/annotate.go [hunk 7]
```diff
@@ -68,8 +70,11 @@ func Run() {
...
+			continue
+		}
+		pairs = append(pairs, fmt.Sprintf("%q:\n%q,", testName, labels))
```
[the remark this answers is not in the document]
Comment (soltysh) [reply openshift-hack/e2e/annotate/annotate.go:77]: Same as above.

File: openshift-hack/e2e/annotate/rules.go (hunk 1 above)
Comment (soltysh) [reply openshift-hack/e2e/annotate/rules.go:117]: Those tests and rules are copied from origin, see [Move k8s-specific rules to our fork](https://github.com/openshift/origin/pull/27570/commits/a29317caa522e29749dba899997e4ef03f2fd71f) commit in origin PR, previously we had a mix of rules in here and in origin. This PR splits the rules so they live where the test is.
